src/library.js

```diff
@@ -967,63 +967,78 @@ addToLibrary({
 
     // reduce number of matchers
     var EQUIVALENT_MATCHERS = {
-      '%A':  '%a',
-      '%B':  '%b',
-      '%c':  '%a %b %d %H:%M:%S %Y',
-      '%D':  '%m\\/%d\\/%y',
-      '%e':  '%d',
-      '%F':  '%Y-%m-%d',
-      '%h':  '%b',
-      '%R':  '%H\\:%M',
-      '%r':  '%I\\:%M\\:%S\\s%p',
-      '%T':  '%H\\:%M\\:%S',
-      '%x':  '%m\\/%d\\/(?:%y|%Y)',
-      '%X':  '%H\\:%M\\:%S'
+      'A':  '%a',
+      'B':  '%b',
+      'c':  '%a %b %d %H:%M:%S %Y',
+      'D':  '%m\\/%d\\/%y',
+      'e':  '%d',
+      'F':  '%Y-%m-%d',
+      'h':  '%b',
+      'R':  '%H\\:%M',
+      'r':  '%I\\:%M\\:%S\\s%p',
+      'T':  '%H\\:%M\\:%S',
+      'x':  '%m\\/%d\\/(?:%y|%Y)',
+      'X':  '%H\\:%M\\:%S'
     };
-    for (var matcher in EQUIVALENT_MATCHERS) {
-      pattern = pattern.replace(matcher, EQUIVALENT_MATCHERS[matcher]);
-    }
```

Comment:
Why change the way EQUIVALENT_MATCHERS work?  Why not keep these three lines and avoid needing the new EQUIVALENT_MATCHERS code below?

Reply:
Because that existing code has a bug. If you put e.g. `%%D`, it gets converted to '%%m/%d/%y`, whereas it should be left as matching a literal % followed by a capital D. I added a test case for this bug, try it in the previous emscripten and it will fail. Doing everything character by character is just way less fragile than repeated find calls.

Reply:
Would a regex help here? So you'd do

```js
var pattern_out =
  str
  .replace(/%(.)/g, (m, c) => EQUIVALENT_MATCHERS[c] || m)
  .replace(/%(.)/g, (_, c) => {
    let pat = DATE_PATTERNS[c];
    return pat ? `(${pat})` : c;
  });
```

Something like this should handle `%%` pairs correctly, and thanks to splitting into 2 passes would be simpler to follow than a manual loop.

Reply:
(also see comment below on handling special regex chars, that applies to all 3 possible implementations)

Reply:
Brilliant. That sorts it with a small addition to make the capture list.